File: src/main/java/leetcode/com/topk/TopkFrequentElements.java
package leetcode.com.topk;

import org.apache.hadoop.yarn.webapp.hamlet.Hamlet;

import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

public class TopkFrequentElements {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();

        for (int num : nums) {
            Integer i = map.get(num);
            if (i != null) {
                map.put(num, i + 1);
            } else {
                map.put(num, 1);
            }
        }

        PriorityQueue<Eleme> queue = new PriorityQueue<>(
                new Comparator<Eleme>() {
                    @Override
                    public int compare(Eleme o1, Eleme o2) {
                        return o2.getCnt() - o1.getCnt();
                    }
                }
        );


        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            Eleme elem = new Eleme(entry.getKey(), entry.getValue());
            queue.offer(elem);
        }


        int[] result = new int[k];

        for (int i = 0; i < k; i++) {
            int e = queue.poll().getE();
            result[i] = e;
        }


        return result;


    }
}

class Eleme {
    private int e;
    private int cnt;

    public Eleme(int e, int cnt) {
        this.e = e;
        this.cnt = cnt;
    }

    public int getE() {
        return e;
    }

    public void setE(int e) {
        this.e = e;
    }

    public int getCnt() {
        return cnt;
    }

    public void setCnt(int cnt) {
        this.cnt = cnt;
    }
}

File: GoProject/leetcode/sort/top-k-frequent-elements.go
package main

import "fmt"

func topKFrequent(nums []int, k int) []int {
	m := make(map[int]int)

	if len(nums) <= k {
		return nums
	}

	l := 0
	for _, num := range nums {
		i, ok := m[num]
		if ok {
			m[num] = i + 1
			if i+1 > l {
				l = i + 1
			}
		} else {
			m[num] = 1
			if l == 0 {
				l = 1
			}
		}
	}

	ints := make([]int, l+1)

	fmt.Println(m)
	for index, cnt := range m {
		ints[cnt] = index
	}

	fmt.Println(ints)
	i := l
	j := 0
	result := []int{}

	for i >= 0 && j < k {
		fmt.Println(i)
		if ints[i] > 0 {
			result = append(result, ints[i])
			j++
		}
		i--
	}

	return result
}

func tmp() {
	a := []int{1, 2, 3, 4, 5}

	for i, _ := range a {
		tmp := make([]int, 0)
		tmp = append(tmp,a[0:i]...)
		tmp = append(tmp,a[i+1:]...)
		fmt.Println(tmp)
	}

}

func main() {
	fmt.Println(topKFrequent([]int{1, 1, 1, 2, 2, 3}, 2))
	tmp()
}
